### PR TITLE
Make SketchJoin ordered serialization aware

### PR DESCRIPTION
**Pull request by MansurAshraf**
Fixes https://github.com/twitter/scalding/issues/1315

### CODE REVIEW

File: scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala [hunk 1]
```diff
@@ -121,6 +129,17 @@ case class SketchJoined[K: Ordering, V, V2, R](left: Sketched[K, V],
...
+  private implicit def intKeyOrd: Ordering[(Int, K)] = {
+    val kord = implicitly[Ordering[K]]
+
+    kord match {
+      case kos: OrderedSerialization[_] => new OrderedSerialization2(ordSer[Int], kos.asInstanceOf[OrderedSerialization[K]])
+      case _ => Ordering.Tuple2[Int, K]
+    }
+  }
+
+  private implicit def ordSer[T]: OrderedSerialization[T] = macro OrderedSerializationProviderImpl[T]
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: let's move this to an object in scalding-serialization. We keep putting this code where we need it, which is needless duplication.

Comment (MansurAshraf) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: @johnynek Cant move it to scalding-serialization as `OrderedSerializationProviderImpl` is in scalding-serialization-macros which depends on scalding-serialization. Cant move it to scalding-serialization-macros either as macro implementation and its usage cannot reside in same module.

**Comment by MansurAshraf** [comment]
BTW we are getting permgen errors again.

File: project/Build.scala [hunk 2]
```diff
@@ -274,7 +274,7 @@ object ScaldingBuild extends Build {
       "org.slf4j" % "slf4j-api" % slf4jVersion,
       "org.slf4j" % "slf4j-log4j12" % slf4jVersion % "provided"
     )
-  ).dependsOn(scaldingArgs, scaldingDate, scaldingSerialization, maple)
+  ).dependsOn(scaldingArgs, scaldingDate, scaldingSerialization, maple, scaldingSerialization, scaldingSerializationMacros)
```
Comment (ianoc) [inline project/Build.scala]: scaldingSerialization was already here?

File: scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala [hunk 3]
```diff
@@ -47,12 +51,15 @@ object Sketched {
...
-case class Sketched[K, V](pipe: TypedPipe[(K, V)],
+case class Sketched[K, V](
+  pipe: TypedPipe[(K, V)],
   numReducers: Int,
   delta: Double,
   eps: Double,
-  seed: Int)(implicit serialization: K => Array[Byte],
-    ordering: Ordering[K])
+  seed: Int
+)(implicit
```
Comment (ianoc) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: These look like formatting changes only? should revert

File: scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala [hunk 4]
```diff
@@ -88,7 +95,8 @@ case class Sketched[K, V](pipe: TypedPipe[(K, V)],
   def leftJoin[V2](right: TypedPipe[(K, V2)]) = cogroup(right)(Joiner.hashLeft2)
 }
 
-case class SketchJoined[K: Ordering, V, V2, R](left: Sketched[K, V],
+case class SketchJoined[K: Ordering, V, V2, R](
```
Comment (ianoc) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: revert formatting change

File: scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala (hunk 1 above)
Comment (ianoc) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: You can put it in scalding-serialization-macros. You can't call it there, but you can define it there, we have it in most of our macro modules.

Comment (MansurAshraf) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: I am calling it in intKeyOrd method



    _____________________________

Comment (ianoc) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: right... you can put this method in scalding-serialization-macros and call it from scalding-core fine though

**Comment by MansurAshraf** [comment]
@johnynek Travis finally turned green. Can we merge this?

**Pull request merged**